### PR TITLE
Add: Implements fasthttp framework with MySQL database.

### PR DESCRIPTION
**Pull request by JammUtkarsh**
Closes [#421](https://github.com/keploy/keploy/issues/421)

### CODE REVIEW

**Comment by JammUtkarsh** [comment]
Please review the codebase, since keploy test is failing but the app seems to be working just fine.

Tasks to be done be
- [x] Implement fasthttp.
- [x] Implement MySQL.
- [ ] API screenshots to be added.
- [ ] Any chores after review.

**Review by nehagup — COMMENTED** [review]
@JammUtkarsh you need to integrate keploy and generate some test cases for this sample app, only then this PR would be accepted.

**Comment by JammUtkarsh** [comment]
I'll like to inform the maintainer that [#178](https://github.com/keploy/go-sdk/issues/178) is blocking this PR, since `keploy/mocks` are not being generated.

As soon as it gets fixed, I'll commit the new updates with tests.

File: fasthttp-mysql/README.md
Comment (SanskritiHarmukh) [inline fasthttp-mysql/README.md:110]: Hi @JammUtkarsh I cannot see any mock generation happening when I tried running the app locally.

Screenshot from 2023-03-23 12-47-15

Screenshot from 2023-03-23 12-48-25

File: fasthttp-mysql/README.md
Comment (SanskritiHarmukh) [inline fasthttp-mysql/README.md:54]: Here, it should be `cd fasthttp-mysql`

File: fasthttp-mysql/README.md
Comment (JammUtkarsh) [reply fasthttp-mysql/README.md:110]: Yes, due to a bug in fasthttp integration as mentioned [above](https://github.com/keploy/samples-go/pull/43#issuecomment-1479452497), the mocks are not being generated.

**Comment by JammUtkarsh** [comment]
When trying to run `export KEPLOY_MODE=test` and `go test -v -coverpkg=./... -covermode=atomic  ./...` I am getting the following output:
```bash keploy                                                                                                                                                 9:13:20

       ▓██▓▄
    ▓▓▓▓██▓█▓▄
     ████████▓▒
          ▀▓▓███▄      ▄▄   ▄               ▌
         ▄▌▌▓▓████▄    ██ ▓█▀  ▄▌▀▄  ▓▓▌▄   ▓█  ▄▌▓▓▌▄ ▌▌   ▓
       ▓█████████▌▓▓   ██▓█▄  ▓█▄▓▓ ▐█▌  ██ ▓█  █▌  ██  █▌ █▓
      ▓▓▓▓▀▀▀▀▓▓▓▓▓▓▌  ██  █▓  ▓▌▄▄ ▐█▓▄▓█▀ █▓█ ▀█▄▄█▀   █▓█
       ▓▌                           ▐█▌                   █▌
        ▓

keploy 0.8.1

.2023-03-27T09:13:39.782+0530	INFO	server/server.go:215	keploy started at port 6789
2023-03-27T09:13:42.548+0530	INFO	testCase/testCase.go:128	no testcases found in /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/tests directory.
2023-03-27T09:13:42.562+0530	INFO	testCase/testCase.go:228
💾 Recorded testcase with name: test-1 in yaml file at path: /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/tests

2023-03-27T09:13:42.564+0530	INFO	testCase/testCase.go:238
💾 Recorded mocks for testcase with name: test-1 at path: /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/mocks

2023-03-27T09:14:14.660+0530	INFO	testCase/testCase.go:228
💾 Recorded testcase with name: test-2 in yaml file at path: /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/tests

2023-03-27T09:14:14.661+0530	INFO	testCase/testCase.go:238
💾 Recorded mocks for testcase with name: test-2 at path: /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/mocks

2023-03-27T09:14:30.059+0530	INFO	testCase/testCase.go:228
💾 Recorded testcase with name: test-3 in yaml file at path: /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/tests

2023-03-27T09:14:30.060+0530	INFO	testCase/testCase.go:238
💾 Recorded mocks for testcase with name: test-3 at path: /Users/utc/code/openSource/samples-go/fasthttp-mysql/keploy/mocks


 <=========================================>
  TESTRUN STARTED with id: "d2db0330-5b69-43c4-99c5-8a11f587c759"
	For App: "fasthttp-mysql"
	Total tests: 0
 <=========================================>

2023-03-27T09:15:30.919+0530	ERROR	regression/regression.go:90	found no test results for test report with id: d2db0330-5b69-43c4-99c5-8a11f587c759
go.keploy.io/server/pkg/service/regression.(*Regression).stopTestRun
	/Users/runner/work/keploy/keploy/pkg/service/regression/regression.go:90
go.keploy.io/server/pkg/service/regression.(*Regression).PutTest
	/Users/runner/work/keploy/keploy/pkg/service/regression/regression.go:898
go.keploy.io/server/http/regression.(*regression).End
	/Users/runner/work/keploy/keploy/http/regression/regression.go:69
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/chi.(*Mux).routeHTTP
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:436
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/chi.(*Mux).ServeHTTP
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:70
github.com/go-chi/chi.(*Mux).Mount.func1
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:311
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/chi.(*Mux).routeHTTP
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:436
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/chi.(*Mux).ServeHTTP
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:70
github.com/go-chi/chi.(*Mux).Mount.func1
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:311
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/chi.(*Mux).routeHTTP
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:436
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/cors.(*Cors).Handler.func1
	/Users/runner/go/pkg/mod/github.com/go-chi/cors@v1.2.0/cors.go:228
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/keploy/go-sdk/integrations/kchi.(*chiV5).Next
	/Users/runner/go/pkg/mod/github.com/keploy/go-sdk@v0.7.9/integrations/kchi/chi-v5.go:64
github.com/keploy/go-sdk/keploy.Middleware
	/Users/runner/go/pkg/mod/github.com/keploy/go-sdk@v0.7.9/keploy/router.go:34
github.com/keploy/go-sdk/integrations/kchi.ChiMiddlewareV5.func1.1
	/Users/runner/go/pkg/mod/github.com/keploy/go-sdk@v0.7.9/integrations/kchi/chi-v5.go:13
net/http.HandlerFunc.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2047
github.com/go-chi/chi.(*Mux).ServeHTTP
	/Users/runner/go/pkg/mod/github.com/go-chi/chi@v1.5.4/mux.go:87
net/http.serverHandler.ServeHTTP
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:2879
net/http.(*conn).serve
	/Users/runner/hostedtoolcache/go/1.17.13/x64/src/net/http/server.go:1930

 <=========================================>
  TESTRUN SUMMARY. For testrun with id: "d2db0330-5b69-43c4-99c5-8a11f587c759"
	Total tests: 0
	Total test passed: 0
	Total test failed: 0
 <=========================================>

^C
```

A visual way of showing the above info is:
main.go

handlers.go

**Comment by aerowisca** [comment]
@JammUtkarsh  Your sample application crashes on my machine. 

**Steps to reproduce**:
1. Run MySQL docker container in the background.
2. Make a POST call to the api.
3. The applications FATALs with this error

```
INFO: Listening on port 8080
INFO: POST /movie
[mysql] 2023/03/27 20:25:20 packets.go:37: unexpected EOF
[mysql] 2023/03/27 20:25:20 packets.go:37: unexpected EOF
[mysql] 2023/03/27 20:25:20 packets.go:37: unexpected EOF
2023/03/27 20:25:20 driver: bad connection
exit status 1
```

I was able to RCA it with the help of this [SO answer](https://stackoverflow.com/questions/39980902/golang-mysql-error-packets-go33-unexpected-eof). Adding the fix worked for me locally. So you might want to reasearch more and incorporate it in your CR.

**Comment by aerowisca** [comment]
Also, I tried replicating the issue with reports that you mentioned, but wasn't able to. It's working fine for me. Can you try the repro steps again?

**Comment by JammUtkarsh** [comment]
> Also, I tried replicating the issue with reports that you mentioned, but wasn't able to. It's working fine for me. Can you try the repro steps again?
@aerowisca

I did replicate the issue and it boils down to this [piece of code](https://github.com/JammUtkarsh/samples-go/blob/fasthttp-mysql/fasthttp-mysql/handlers.go#L51-L58). You may check it once if there is anything wrong with it. I don't know what's wrong there. I have also added more info in issue #510.

**Comment by JammUtkarsh** [comment]
@SanskritiHarmukh
The PR can finally be merged. This issue #510 was fixed in [#183](https://github.com/keploy/go-sdk/pull/183) with go-sdk version bump from 0.8.3 to 0.8.4.

Screenshot 2023-04-02 at 20 20 00
Screenshot 2023-04-02 at 20 20 20

**Comment by SanskritiHarmukh** [comment]
Hi @JammUtkarsh good work. I tried the app locally, works fine for me.

Screenshot from 2023-04-03 13-05-04

**Can you also include a section where you introduce a bug in code and then make API calls and do test run to check if keploy can catch it?**

**Review by nehagup — APPROVED** [review]

**Comment by JammUtkarsh** [comment]
Here is a failing test which keploy catches.

I changed the data type from `int` to `string`

Screenshot 2023-04-03 at 13 25 20

It gives the following log in Keploy
```bash
Testrun failed for testcase with id: "test-1"
Test Result:
	Input Http Request: models.HttpReq{
  Method:     "POST",
  ProtoMajor: 1,
  ProtoMinor: 1,
  URL:        "/movie",
  URLParams:  map[string]string{},
  Header:     http.Header{
    "Accept": []string{
      "*/*",
    },
    "Content-Length": []string{
      "85",
    },
    "Content-Type": []string{
      "application/json",
    },
    "Host": []string{
      "localhost:8080",
    },
    "User-Agent": []string{
      "curl/8.0.1",
    },
  },
  Body:   "{\n    \"title\":\"Everything Everywhere All At Once\",\n    \"year\":2022,\n    \"rating\":10\n}",
  Binary: "",
  Form:   []models.FormData(nil),
}

	Expected Response: models.HttpResp{
  StatusCode: 200,
  Header:     http.Header{
    "Content-Type": []string{
      "text/plain; charset=utf-8",
    },
    "Server": []string{
      "fasthttp",
    },
  },
  Body:          "{\"id\":0,\"title\":\"Everything Everywhere All At Once\",\"year\":2022,\"rating\":10}",
  StatusMessage: "",
  ProtoMajor:    0,
  ProtoMinor:    0,
  Binary:        "",
}

	Actual Response: models.HttpResp{
  StatusCode: 400,
  Header:     http.Header{
    "Content-Type": []string{
      "text/plain; charset=utf-8",
    },
    "Server": []string{
      "fasthttp",
    },
  },
  Body:          "Invalid JSON",
  StatusMessage: "",
  ProtoMajor:    0,
  ProtoMinor:    0,
  Binary:        "",
}

DIFF:
	Expected StatusCode: 200
	Actual StatusCode: 400

	Response body: {
{
			Expected value: "{\"id\":0,\"title\":\"Everything Everywhere All At Once\",\"year\":2022,\"rating\":10}"
			Actual value: "Invalid JSON"
		}
--------------------------------------------------------------------

Testrun failed for testcase with id: "test-3"
Test Result:
	Input Http Request: models.HttpReq{
  Method:     "GET",
  ProtoMajor: 1,
  ProtoMinor: 1,
  URL:        "/movies",
  URLParams:  map[string]string{},
  Header:     http.Header{
    "Accept": []string{
      "*/*",
    },
    "Host": []string{
      "localhost:8080",
    },
    "User-Agent": []string{
      "curl/8.0.1",
    },
  },
  Body:   "",
  Binary: "",
  Form:   []models.FormData(nil),
}

	Expected Response: models.HttpResp{
  StatusCode: 200,
  Header:     http.Header{
    "Content-Type": []string{
      "text/plain; charset=utf-8",
    },
    "Server": []string{
      "fasthttp",
    },
  },
  Body:          "[{\"id\":1,\"title\":\"The Godfather\",\"year\":1972,\"rating\":9},{\"id\":2,\"title\":\"The Shawshank Redemption\",\"year\":1994,\"rating\":9},{\"id\":3,\"title\":\"The Dark Knight\",\"year\":2008,\"rating\":7},{\"id\":4,\"title\":\"The Godfather: Part II\",\"year\":1974,\"rating\":9},{\"id\":5,\"title\":\"The Lord of the Rings: The Return of the King\",\"year\":2003,\"rating\":6},{\"id\":6,\"title\":\"Pulp Fiction\",\"year\":1994,\"rating\":8},{\"id\":7,\"title\":\"Schindler's List\",\"year\":1993,\"rating\":8},{\"id\":8,\"title\":\"The Good, the Bad and the Ugly\",\"year\":1966,\"rating\":8},{\"id\":9,\"title\":\"The Lord of the Rings: The Fellowship of the Ring\",\"year\":2001,\"rating\":6},{\"id\":10,\"title\":\"The Lord of the Rings: The Two Towers\",\"year\":2002,\"rating\":6},{\"id\":11,\"title\":\"Fight Club\",\"year\":1999,\"rating\":8},{\"id\":12,\"title\":\"The Matrix\",\"year\":1999,\"rating\":8},{\"id\":13,\"title\":\"Forrest Gump\",\"year\":1994,\"rating\":8},{\"id\":14,\"title\":\"Inception\",\"year\":2010,\"rating\":8},{\"id\":15,\"title\":\"The Dark Knight Rises\",\"year\":2012,\"rating\":7},{\"id\":16,\"title\":\"Everything Everywhere All At Once\",\"year\":2022,\"rating\":10}]",
  StatusMessage: "",
  ProtoMajor:    0,
  ProtoMinor:    0,
  Binary:        "",
}

	Actual Response: models.HttpResp{
  StatusCode: 200,
  Header:     http.Header{
    "Content-Type": []string{
      "text/plain; charset=utf-8",
    },
    "Server": []string{
      "fasthttp",
    },
  },
  Body:          "[{\"id\":1,\"title\":\"The Godfather\",\"year\":\"1972\",\"rating\":9},{\"id\":2,\"title\":\"The Shawshank Redemption\",\"year\":\"1994\",\"rating\":9},{\"id\":3,\"title\":\"The Dark Knight\",\"year\":\"2008\",\"rating\":7},{\"id\":4,\"title\":\"The Godfather: Part II\",\"year\":\"1974\",\"rating\":9},{\"id\":5,\"title\":\"The Lord of the Rings: The Return of the King\",\"year\":\"2003\",\"rating\":6},{\"id\":6,\"title\":\"Pulp Fiction\",\"year\":\"1994\",\"rating\":8},{\"id\":7,\"title\":\"Schindler's List\",\"year\":\"1993\",\"rating\":8},{\"id\":8,\"title\":\"The Good, the Bad and the Ugly\",\"year\":\"1966\",\"rating\":8},{\"id\":9,\"title\":\"The Lord of the Rings: The Fellowship of the Ring\",\"year\":\"2001\",\"rating\":6},{\"id\":10,\"title\":\"The Lord of the Rings: The Two Towers\",\"year\":\"2002\",\"rating\":6},{\"id\":11,\"title\":\"Fight Club\",\"year\":\"1999\",\"rating\":8},{\"id\":12,\"title\":\"The Matrix\",\"year\":\"1999\",\"rating\":8},{\"id\":13,\"title\":\"Forrest Gump\",\"year\":\"1994\",\"rating\":8},{\"id\":14,\"title\":\"Inception\",\"year\":\"2010\",\"rating\":8},{\"id\":15,\"title\":\"The Dark Knight Rises\",\"year\":\"2012\",\"rating\":7},{\"id\":16,\"title\":\"Everything Everywhere All At Once\",\"year\":\"2022\",\"rating\":10}]",
  StatusMessage: "",
  ProtoMajor:    0,
  ProtoMinor:    0,
  Binary:        "",
}

DIFF:
	Response body: {
		"": {
			Expected value: "[{\"id\":1,\"rating\":9,\"title\":\"The Godfather\",\"year\":1972},{\"id\":2,\"rating\":9,\"title\":\"The Shawshank Redemption\",\"year\":1994},{\"id\":3,\"rating\":7,\"title\":\"The Dark Knight\",\"year\":2008},{\"id\":4,\"rating\":9,\"title\":\"The Godfather: Part II\",\"year\":1974},{\"id\":5,\"rating\":6,\"title\":\"The Lord of the Rings: The Return of the King\",\"year\":2003},{\"id\":6,\"rating\":8,\"title\":\"Pulp Fiction\",\"year\":1994},{\"id\":7,\"rating\":8,\"title\":\"Schindler's List\",\"year\":1993},{\"id\":8,\"rating\":8,\"title\":\"The Good, the Bad and the Ugly\",\"year\":1966},{\"id\":9,\"rating\":6,\"title\":\"The Lord of the Rings: The Fellowship of the Ring\",\"year\":2001},{\"id\":10,\"rating\":6,\"title\":\"The Lord of the Rings: The Two Towers\",\"year\":2002},{\"id\":11,\"rating\":8,\"title\":\"Fight Club\",\"year\":1999},{\"id\":12,\"rating\":8,\"title\":\"The Matrix\",\"year\":1999},{\"id\":13,\"rating\":8,\"title\":\"Forrest Gump\",\"year\":1994},{\"id\":14,\"rating\":8,\"title\":\"Inception\",\"year\":2010},{\"id\":15,\"rating\":7,\"title\":\"The Dark Knight Rises\",\"year\":2012},{\"id\":16,\"rating\":10,\"title\":\"Everything Everywhere All At Once\",\"year\":2022}]"
			Actual value: "[{\"id\":1,\"rating\":9,\"title\":\"The Godfather\",\"year\":\"1972\"},{\"id\":2,\"rating\":9,\"title\":\"The Shawshank Redemption\",\"year\":\"1994\"},{\"id\":3,\"rating\":7,\"title\":\"The Dark Knight\",\"year\":\"2008\"},{\"id\":4,\"rating\":9,\"title\":\"The Godfather: Part II\",\"year\":\"1974\"},{\"id\":5,\"rating\":6,\"title\":\"The Lord of the Rings: The Return of the King\",\"year\":\"2003\"},{\"id\":6,\"rating\":8,\"title\":\"Pulp Fiction\",\"year\":\"1994\"},{\"id\":7,\"rating\":8,\"title\":\"Schindler's List\",\"year\":\"1993\"},{\"id\":8,\"rating\":8,\"title\":\"The Good, the Bad and the Ugly\",\"year\":\"1966\"},{\"id\":9,\"rating\":6,\"title\":\"The Lord of the Rings: The Fellowship of the Ring\",\"year\":\"2001\"},{\"id\":10,\"rating\":6,\"title\":\"The Lord of the Rings: The Two Towers\",\"year\":\"2002\"},{\"id\":11,\"rating\":8,\"title\":\"Fight Club\",\"year\":\"1999\"},{\"id\":12,\"rating\":8,\"title\":\"The Matrix\",\"year\":\"1999\"},{\"id\":13,\"rating\":8,\"title\":\"Forrest Gump\",\"year\":\"1994\"},{\"id\":14,\"rating\":8,\"title\":\"Inception\",\"year\":\"2010\"},{\"id\":15,\"rating\":7,\"title\":\"The Dark Knight Rises\",\"year\":\"2012\"},{\"id\":16,\"rating\":10,\"title\":\"Everything Everywhere All At Once\",\"year\":\"2022\"}]"
		}
	}
--------------------------------------------------------------------

Testrun failed for testcase with id: "test-2"
Test Result:
	Input Http Request: models.HttpReq{
  Method:     "GET",
  ProtoMajor: 1,
  ProtoMinor: 1,
  URL:        "/movie",
  URLParams:  map[string]string{},
  Header:     http.Header{
    "Accept": []string{
      "*/*",
    },
    "Host": []string{
      "localhost:8080",
    },
    "User-Agent": []string{
      "curl/8.0.1",
    },
  },
  Body:   "",
  Binary: "",
  Form:   []models.FormData(nil),
}

	Expected Response: models.HttpResp{
  StatusCode: 200,
  Header:     http.Header{
    "Content-Type": []string{
      "text/plain; charset=utf-8",
    },
    "Server": []string{
      "fasthttp",
    },
  },
  Body:          "{\"id\":16,\"title\":\"Everything Everywhere All At Once\",\"year\":2022,\"rating\":10}",
  StatusMessage: "",
  ProtoMajor:    0,
  ProtoMinor:    0,
  Binary:        "",
}

	Actual Response: models.HttpResp{
  StatusCode: 200,
  Header:     http.Header{
    "Content-Type": []string{
      "text/plain; charset=utf-8",
    },
    "Server": []string{
      "fasthttp",
    },
  },
  Body:          "{\"id\":16,\"title\":\"Everything Everywhere All At Once\",\"year\":\"2022\",\"rating\":10}",
  StatusMessage: "",
  ProtoMajor:    0,
  ProtoMinor:    0,
  Binary:        "",
}

DIFF:
	Response body: {
		"": {
			Expected value: "{\"id\":16,\"rating\":10,\"title\":\"Everything Everywhere All At Once\",\"year\":2022}"
			Actual value: "{\"id\":16,\"rating\":10,\"title\":\"Everything Everywhere All At Once\",\"year\":\"2022\"}"
		}
	}
```

The `DIFF:` here in keploy shows the actual change which you might be looking for @SanskritiHarmukh.

**Comment by SanskritiHarmukh** [comment]
Cool

**Review by SanskritiHarmukh — APPROVED** [review]

**Comment by JammUtkarsh** [comment]
Hey @nehagup, if everything looks good, can I know why does it have a `don't merge` tag?

The reason I ask this is because, this question was asked to me in an interview and I didn
t have an answer there. I was hoping you help me with it.

Regards,
Utkarsh

**Comment by nehagup** [comment]
@JammUtkarsh it's because it was a GSoC task and we were not allowed to merge request from one participant to avoid biases.

**Comment by nehagup** [comment]
@JammUtkarsh closing this for now since keploy v2 is launched and this PR has integration with keploy v1. Please feel free to modify the integration according to Keploy v2 and make a PR, will definitely merge.

**Comment by JammUtkarsh** [comment]
Sure, that looks promising, I'll definitely do that!

Thank you for resolving the doubt.

**Pull request closed without merging**
